os/mm/mm_heap/mm_mallinfo.c

```diff
@@ -113,36 +116,11 @@ int mm_mallinfo(FAR struct mm_heap_s *heap, FAR struct mallinfo *info)
 
 		mm_takesemaphore(heap);
 
-		prev = NULL;
-		node = (struct mm_freenode_s *)heap->mm_heapstart[region];
-		next = (struct mm_freenode_s *)((char *)node + node->size);
-
-		for (; node < (struct mm_freenode_s *)heap->mm_heapend[region]; prev = node, node = next, next = (struct mm_freenode_s *)((char *)next + next->size)) {
-			mvdbg("region=%d node=%p size=%u preceding=%u (%c)\n", region, node, node->size, (node->preceding & ~MM_ALLOC_BIT), (node->preceding & MM_ALLOC_BIT) ? 'A' : 'F');
-
-			if ((prev && prev->size != (node->preceding & ~MM_ALLOC_BIT)) ||
-				(node->size != (next->preceding & ~MM_ALLOC_BIT)) ||
-				(!(node->preceding & MM_ALLOC_BIT) &&
-				((node->blink && node->blink->flink != node) ||
-				(node->flink && node->flink->blink != node)))
-			) {
-				mdbg("ERROR: Heap node corruption detected from one of following nodes\n");
-				if (prev) {
-					mdbg("Previous node addr = 0x%08x size = %u preceding size = %u\n", prev, prev->size, prev->preceding & ~MM_ALLOC_BIT);
-				}
-				mdbg("Current  node addr = 0x%08x size = %u preceding size = %u\n", node, node->size, node->preceding & ~MM_ALLOC_BIT);
-
-#ifdef CONFIG_DEBUG_MM_HEAPINFO
-				if (prev) {
-					mdbg("Previous node owner pid = %u, alloc_call_addr = 0x%08x\n", prev->pid, prev->alloc_call_addr);
-				}
-				mdbg("Current  node owner pid = %u, alloc_call_addr = 0x%08x\n", node->pid, node->alloc_call_addr);
-#endif
-				ASSERT(0);
```

Comment:
Could you let me know why you remove this from mm_mallinfo?
If because this is duplicated, then we should revert removing previous assert at line 138.
```c
mvdbg("region=%d node=%p heapend=%p\n", region, node, heap->mm_heapend[region]);
DEBUGASSERT(node == heap->mm_heapend[region]);
uordblks += SIZEOF_MM_ALLOCNODE;	/* account for the tail node */
```

When you add debugassert, I want to know your opinion for debugassert and assert for that.

Reply:
Since we are checking for heap corruption, i think we dont need this check. 
A condition of last node in the loop not matching the heapend will occur only if one of the nodes in the middle is corrupted. 
But, i think we could instead add DEBUGASSERT in the heap corruption check api so that during debugging, we will halt if heap corruption is seen. But, in the release binary, we will let the execution to continue even with heap corruption. 
Please let me know if its ok.

Reply:
Even we can know heap corruption at mm_mallinfo, we can't know it at that time.
We will get assert some time later, or it could cause some strange symptom without assert.
mm_mallinfo is not mandatory API so that adding one conditional does not affect performance degradation. Then I think why not.

Reply:
What I meant is that, if there is a corruption in any node, it will assert at line 218. 
`DEBUGASSERT(mm_check_heap_corruption(heap) == OK)`

There will be no need to go till end of heap. 
But, if the for loop actually goes till end of loop without asserting at line 218, then it means there is no corruption. 
So, this check is redundant. 

`DEBUGASSERT(node == heap->mm_heapend[region]);`

Reply:
Oh I didn't find line 217. But in my opinion, adding `ASSERT(node == heap->mm_heapend[region]);` is better instead adding line 217.
1. This is working only when both config_debug and config_debug_mm_heapinfo are enabled. I want to check heap corruption anytime. (I mean it should work without any conditional).
2. This causes walking entire nodes twice. If you want to add this here, then you should merge this and second for loop.

The number of checking entire node is
|    | Adding one line at end | Current |
|---------------------------|------|-----|
| No Memory Corrupted | 1 | 2 |
| Some Memory Corrupted | 2 | 3 |

Reply:
> This is working only when both config_debug and config_debug_mm_heapinfo are enabled. I want to check heap corruption anytime. (I mean it should work without any conditional).

Sorry, I missed removing the heapinfo condition. Also, I forgot about DEBUGASSERT. I modified it as below. So, it will always check corruption:

`        if (mm_check_heap_corruption(heap) != OK) {`
`                DEBUGASSERT(0);`
`        }`

> This causes walking entire nodes twice. If you want to add this here, then you should merge this and second for loop.

I understand your concern. But, I thought it would be better to separate it into a different API which can be easily called from assert handler, heapinfo and any other location as we require. If I merge the new API into mallinfo, then we will have problem of recursive assertion:

`mallinfo -> corruption detected -> debug assert -> assert handler -> mallinfo`

Even if we dont use debugassert for corruption, its possible that corrupted node might cause assertion in mallinfo and then assert handler will again call mallinfo recursively. So, I feel its better to keep it as a separate API.

Reply:
I completely agree about separate API. But I don't want to call it from mm_mallinfo.
In mm_mallinfo, just check simple one conditional, `ASSERT(node == heap->mm_heapend[region]);`.
If there is corruption, then it will call assert and assert calls your new API.
In this PR, what I want is removing calling `mm_check_heap_corruption` in mm_mallinfo and adding above at end of the function.

Reply:
Done.